easybuild/framework/extension.py

```diff
@@ -172,29 +172,93 @@ def version(self):
         return self.ext.get('version', None)
 
     def prerun(self):
+        """
+        [DEPRECATED][6.0] Stuff to do before installing a extension.
+        """
+        # Deprecation warning triggered by Extension.install_extension_substep()
+        self.pre_install_extension()
+
+    def pre_install_extension(self):
         """
         Stuff to do before installing a extension.
         """
         pass
 
     def run(self, *args, **kwargs):
+        """
+        [DEPRECATED][6.0] Actual installation of an extension.
+        """
+        # Deprecation warning triggered by Extension.install_extension_substep()
+        self.install_extension(*args, **kwargs)
+
+    def install_extension(self, *args, **kwargs):
         """
         Actual installation of an extension.
         """
         pass
 
     def run_async(self, *args, **kwargs):
+        """
+        [DEPRECATED][6.0] Asynchronous installation of an extension.
+        """
+        # Deprecation warning triggered by Extension.install_extension_substep()
+        self.install_extension_async(*args, **kwargs)
+
+    def install_extension_async(self, *args, **kwargs):
         """
         Asynchronous installation of an extension.
         """
         raise NotImplementedError
 
     def postrun(self):
+        """
+        [DEPRECATED][6.0] Stuff to do after installing a extension.
+        """
+        # Deprecation warning triggered by Extension.install_extension_substep()
+        self.post_install_extension()
+
+    def post_install_extension(self):
         """
         Stuff to do after installing a extension.
         """
         self.master.run_post_install_commands(commands=self.cfg.get('postinstallcmds', []))
 
+    def install_extension_substep(self, substep):
+        """
+        Carry out extension installation substep allowing use of deprecated
+        methods on those extensions using an older EasyBlock
+        """
+        deprecated = {
+            'pre_install_extension': 'prerun',
+            'install_extension': 'run',
+            'install_extension_async': 'run_async',
+            'post_install_extension': 'postrun',
+        }
+
+        if substep not in deprecated:
+            raise EasyBuildError("Unknown extension installation substep: %s", substep)
+
+        try:
+            ext_substep = getattr(self, deprecated[substep])
+            parent_obj = super(self.__class__, self)
+            parent_substep = getattr(parent_obj, deprecated[substep])
```

Comment:
@lexming What about the case where the easyblock being used doesn't define (for example) `run` itself, but it derives from another easyblock that does still define it.

Example: an easyblock that derives from a generic easyblock like `PythonPackage`.
If `PythonPackage` still implements `run`, we should still call it, right?

We should only skip calling `run` if the method we obtain is the one from `EasyBlock`?

Reply:
That was indeed a problem. For instance, the `numpy` easyblock does not define any `prerun`, but `PythonPackage` does. So the catch on that was failing because the `prerun` method on numpy is equal to its parent class.

I figured out a simpler more robust method by just using the qualified name of the method. Then it becomes trivial to know what class defines the method of the _active_ EasyBlock. Please try again with latest commits.